PR2

AlexanderSparrow opened the pull request:
Финальное задание 11го спринта. Новый функционал, рефакторинг проекта, логгирование

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Изменена структура проекта: добавлены слои storage, service. Добавлена бизнес логика по управлению пользователями и фильиами (добавить, удалить, и пр)

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлены методы, позволяющие пользователям добавлять друг друга в друзья, получать список общих друзей и лайкать фильмы

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. fix some bugs> добавлена валидация фильма на дату создания при обновлении

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование
----
исправлена ошибка checkstyle

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование
----
исправлена ошибка checkstyle

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,53 +2,49 @@
-    private static final LocalDate startReleaseDate = LocalDate.of(1895, 12, 28);
-    private final Map<Integer, Film> films = new HashMap<>();
-    private int nextId = 1;
+    public FilmController(FilmService filmService) {
```
Можно лучше
Рекомендую для создания конструкторов использовать ломбок. 
Есть аннотации
@NoArgsConstructor создает пустой конструктор
@AllArgsConstructor создает конструктор со всеми полями
@RequiredArgsConstructor создает конструктор для final полей
В нашем случае подходит @RequiredArgsConstructor, но я рекомендую использовать именно эту аннотацию, т.к. в этом случае заставляет разработчика добавлять final, что лучше в плане безопасности, но конечно, где это возможно

AlexanderSparrow replied:
добавил аннотацию @RequiredArgsConstructor, убрал написанный конструктор

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,53 +2,49 @@
-            throw new ValidationException("Дата релиза должна быть не раньше 28 декабря 1895 года");
-        }
+    @GetMapping("/popular")
+    public List<Optional<Film>> getPopularFilms(@RequestParam(defaultValue = "10") int count) {
```
Обязательно
Нужно отдавать List<Film> пользователь ничего не знает про optional

AlexanderSparrow replied:
Доработал метод, добавив проверку непустого Optional<Film>

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,51 @@
+        return filmStorage.updateFilm(film);
+    }
+
+    public void addLike(int filmId, int userId) {
```
Обязательно
При добавлении  и удаление лайков нужно добавить проверку, что такой фильм и пользователь есть

AlexanderSparrow replied:
добавил проверки на существование фильма и юзера

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,51 @@
+@Service
+public class FilmService {
+    private final FilmStorage filmStorage;
+    private final Map<Integer, Set<Integer>> likes = new HashMap<>();
```
Обязательно
Лайки храняться в самом фильме, а не в отдельном хранилище. Иначе у вас могут быть не согласованны данные. Добавьте поле 
private Set<Long> userIds = new HashSet<>();
в film

AlexanderSparrow replied:
Да будет так!:)

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,63 @@
+@Service
+public class UserService {
+    private final UserStorage userStorage;
+    private final Map<Integer, Set<Integer>> userFriends = new HashMap<>();
```
Обязательно
Также как и в фильмах, список друзей храниться в самом пользователи

AlexanderSparrow replied:
Выполнено

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,63 @@
+        return userStorage.updateUser(user);
+    }
+
+    public void addFriend(int userId, int friendId) {
```
Обязательно
При добавлении и удалении друзей нужно проверить, что нам верно переданы id пользователей

AlexanderSparrow replied:
проверки добавлены, исключения обработаны

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,63 @@
+        return userStorage.getAllUsers();
+    }
+
+    public Set<Integer> getUserFriends(int userId) {
```
Обязательно
Нужно также проверить, что такой пользователь есть

AlexanderSparrow replied:
да, учитель:)

kesch9 reviewed (commented):
Нужно подправить реализацию

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование
----
fix some bugs

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование
----
fix some Checkstyle bugs

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование
----
fix some build bugs

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование
----
и снова ЖИРНЫЙ пробел:)

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public void addLike(int filmId, int userId) {
+        if (filmStorage.getFilmById(filmId).isEmpty()) {
```
Обязательно
Эта проверка лишняя, вы проверяете наличие фильма в 33 строке

AlexanderSparrow replied:
Поправил

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,65 @@
+        if (filmStorage.getFilmById(filmId).isEmpty()) {
+            throw new FilmNotFoundException(filmId);
+        }
+        if (userService.getUserById(userId).isEmpty()) {
```
Обязательно
У optional есть метод orElseThrow, поэтому эту запись можно сократить
userService.getUserById(userId).orElseThrow(() -> new UserNotFoundException(userId));
Вы уже используете эту запись сделайте одиннаково везде

AlexanderSparrow replied:
Поправил

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,65 @@
+        if (userService.getUserById(userId).isEmpty()) {
+            throw new UserNotFoundException(userId);
+        }
+        if (userService.getUserById(userId).isEmpty()) {
```
Обязательно
Вы проверяете два раза пользователя и в 40 и в 43 строчке, проверку оставьте одну

AlexanderSparrow replied:
Поправил

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,65 @@
+
+    public List<Film> getPopularFilms(int count) {
+        return filmStorage.getAllFilms().stream()
+                .sorted((film1, film2) -> {
```
Можно лучше
Создайте компоратор для сортировки как переменную
public static final Comparator<Film> FILM_COMPARATOR = Comparator.comparingLong(Film::getRate).reversed();
добавьте методо getRate возвращаюсь количество лайков и тогда можно написать в виде лямбды или не добавляйте этот метод, а используйте ваш вариант

AlexanderSparrow replied:
Реализовал предложенный вариант

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,89 @@
+        return userStorage.getUserById(userId);
+    }
+
+        public Set<Friend> getUserFriends(int userId) {
```
Обязательно
Здесь возвращать нужно не Friend, а самого User

AlexanderSparrow replied:
Исправил

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/model/Friend.java:
```diff
@@ -0,0 +1,12 @@
+@Getter
+@Setter
+@AllArgsConstructor
+public class Friend {
```
Обязательно
Friend не нужен, возвращать нужно User
Александр вы не совсем верно поняли запись в тестах
response.json()[0].id
Это значит, что возвращеном json есть массив, в котором есть объект у которого есть поле id, это не значит, что у него только id.
Здесь можно написать и так
response.json()[0].name
и проверять по имени.
Поэтому нужно отдавать просто User

AlexanderSparrow replied:
Исправил

kesch9 replied:
Сам класс Friend надо удалить он не нужен

kesch9 reviewed (commented):
Нужно еще поработать над заданием

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Добавлено логгирование
----
Устранены замечания по ревью

kesch9 reviewed (approved):
Approve
